jgrapht-core/src/main/java/org/jgrapht/util/SupplierUtil.java

```diff
@@ -56,29 +60,50 @@
      * @return the supplier
      * @param <T> the type of results supplied by this supplier
      */
+    @SuppressWarnings("unchecked")
     public static <T> Supplier<T> createSupplier(Class<? extends T> clazz)
     {
+        // shortcut to use pre-defined constructor method reference based suppliers
+        if (clazz == DefaultEdge.class) {
+            return (Supplier<T>) DEFAULT_EDGE_SUPPLIER;
+        } else if (clazz == DefaultWeightedEdge.class) {
+            return (Supplier<T>) DEFAULT_WEIGHTED_EDGE_SUPPLIER;
+        } else if (clazz == Object.class) {
+            return (Supplier<T>) OBJECT_SUPPLIER;
+        }
+
         try {
             final Constructor<? extends T> constructor = clazz.getDeclaredConstructor();
             if ((!Modifier.isPublic(constructor.getModifiers())
                 || !Modifier.isPublic(constructor.getDeclaringClass().getModifiers()))
                 && !constructor.canAccess(null))
+            {
                 constructor.setAccessible(true);
+            }
             return new ConstructorSupplier<>(constructor);
         } catch (ReflectiveOperationException e) {
-            String error = e.getMessage();
-            return new AlwaysFailingSupplier<T>(error);
+            // Defer throwing an exception to the first time the supplier is called
+            return getThrowingSupplier(e);
         }
     }
 
+    @SuppressWarnings("unchecked")
+    private static <T> Supplier<T> getThrowingSupplier(ReflectiveOperationException e)
+    {
+        String error = e.getMessage();
+        return (Supplier<T> & Serializable) () -> {
+            throw new IllegalStateException(error);
```

Comment:
Here and below, why IllegalStateException?  I don't think it's meant for such a failure.

Reply:
I changed it because from what I know, using a "raw" RuntimeException is considered bad style because it is very generic. And a `IllegalStateException` seems to be the most suitable specific RuntimeException from the Java standard-library because when a `ReflectiveOperationException` is thrown while trying to obtain the constructor it is somehow an illegal state. But maybe an IllegalArgumentException is more suitable here? Or should I create an own, more suitable RuntimeException subclass?

Reply:
I guess a custom exception would be better. Maybe `InvalidSupplierException` ?

Reply:
Since the class of the objects to supply is invalid I think `InvalidSuppliedClassException` is the most precise name, isn't it?

Reply:
Maybe just `SupplierException`?  

@jsichi Could you choose a suitable name?